src/sparse/KokkosSparse_BlockCrsMatrix.hpp

```diff
@@ -0,0 +1,940 @@
+            }
+          }
+        }
+        ++numValid;
+      }
+    } // end for ncol
+    return numValid;
+  }
+
+
+  // replaceValues input:
+  // rowi   is a block-row index
+  // ncol   is number of blocks referenced in cols[] array
+  // cols[] are block colidxs within the block-row to be summed into
+  //        ncol entries
+  // vals[] array containing 'block' of values
+  //        ncol*block_size*block_size entries
+  //        assume vals block is provided in 'LayoutRight' or 'Row Major' format, that is 
+  //        e.g. 2x2 block [ a b ; c d ] provided as flattened 1d array as [a b c d]
+  //        Assume that each block is stored contiguously in vals:
+  //        [a b; c d] [e f; g h] -> [a b c d e f g h]
+  //        If so, then i in [0, ncols) for cols[] 
+  //        maps to i*block_size*block_size in vals[]
+  KOKKOS_INLINE_FUNCTION
+  OrdinalType
+  replaceValues (const OrdinalType rowi,
+                 const OrdinalType cols[],
+                 const OrdinalType ncol,
+                 const ScalarType vals[],
+                 const bool is_sorted = false,
+                 const bool force_atomic = false) const
+  {
+    SparseBlockRowView<BlockCrsMatrix> row_view = this->block_row (rowi);
+    const ordinal_type length = row_view.length;
+    const ordinal_type block_size = this->blockDim();
+
+    ordinal_type numValid = 0; // number of valid local column indices
+
+    for (ordinal_type i = 0; i < ncol; ++i) {
+      // Find offset into values for block-row rowi and colidx cols[i]
+      // cols[i] is the index to match
+      // blk_offset is the offset for block colidx from bptr[rowi] to bptr[rowi + 1] (not global offset)
+      // colidx_ and values_ are already offset to the beginning of blockrow rowi
+      auto blk_offset = row_view.findRelBlockOffset(cols[i], is_sorted);
+      if ( blk_offset != std::numeric_limits<ordinal_type>::max() ) {
+        ordinal_type offset_into_values = blk_offset*block_size;
+        ordinal_type offset_into_vals = i*block_size*block_size; //stride == 1 assumed between elements
+        ordinal_type values_row_stride = block_size*length; // stride to start of next row
+        for ( ordinal_type lrow = 0; lrow < block_size; ++lrow ) {
+          auto local_row_values = row_view.local_row_in_block(blk_offset, lrow); // pointer to start of specified local row within this block
+          for ( ordinal_type lcol = 0; lcol < block_size; ++lcol ) {
+            if (force_atomic) {
+              Kokkos::atomic_assign(&(local_row_values[lcol]), vals[ offset_into_vals + lrow*block_size + lcol ]);
+            }
+            else {
+              local_row_values[lcol] = vals[ offset_into_vals + lrow*block_size + lcol];
+            }
+          }
+        }
+        ++numValid;
+      }
+    } // end for ncol
+    return numValid;
+  }
+
+  //! Attempt to assign the input matrix to \c *this.
+  // Are the CUDA sparse handles needed to be copied here??
```

Comment:
Nothing uses the cuSPARSE handles; please get rid of them.

Reply:
Done :)